Add Dependency Confusion document to upstream docs (#544)
release-engineering/cachito: docs/dependency_confusion.md

define the name of your module as well as the names and versions of its direct dependencies.
Thanks to golang’s unique system of dependency management, this file is enough to ensure
reproducible builds on its own.

However, due to golang versions being tied to git tags, you are technically at the mercy of
repository maintainers. To prevent surprises, whenever go downloads any dependencies, it
saves their checksums in a [go.sum](https://golang.org/ref/mod#go) file. This file is later used to
verify that the content of all direct and indirect dependencies is what you expect it to be. It’s
important to note that once Cachito downloads a dependency from the Internet, it caches it for
future usage. So even if a repository maintainer changes the git tags, Cachito will not detect this
change. 

As long as you make sure to always commit your go.sum file, your Golang module should be safe from
dependency attacks.

The one caveat is Cachito’s own
[dependency replacements](https://github.com/release-engineering/cachito#feature-support) feature.
This feature allows you to replace dependencies with different versions on the fly. Currently, there

> isno way to verify their checksums. Always check that the replacement versions are what you want

Reviewer: ```suggestion
is no way to verify their checksums. Always check that the replacement versions are what you want
```

# Problem Statement

Recent research has shown novel [ways](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) to inject malicious content into applications. The question has been raised as to whether or not using Cachito makes this issue worse. This document is an analysis of each package manager supported by Cachito to address the security concern named *dependency confusion*.

# Dependency Confusion

> To paraphrase the meaning of *[dependency confusion](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610)*: a package manager is tricked into installing a dependency from an official, public, repository when it should’ve installed it from a custom, often private, repository. For example, a company may host their own internal npm registry which hosts internal-only packages. Then, someone on the Internet uploads a package of the same name to the official npm repository. When the application is installed, the package manager chooses the dependency package from the public repository instead of the internal one.

Reviewer 2: ```suggestion
To paraphrase the meaning of *[dependency confusion](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610)*: a package manager is tricked into installing a dependency from an official, public, repository when it should’ve installed it from a custom, often private, repository. For example, a company may host its internal npm registry which hosts internal-only packages. Then, someone on the Internet uploads a package of the same name to the official npm repository. When the application is installed, the package manager chooses the dependency package from the public repository instead of the internal one.
```

Reviewer 2: Or maybe I am wrong :D

Author: You're right, I'll make that change!

## Potential Impact

What happens when Cachito is tricked into downloading a malicious version of one of your dependencies? To Cachito itself, nothing. Cachito never executes your code, or the code of any of your dependencies. What about your build?

> You are likely using Cachito through OSBS. Cachito will serve the malicious dependency to the OSBS build, where you will subsequently use it to build your application container. The malicious behaviour may or may not manifest at build time. In the best case scenario, the build will fail. In the worst case scenario, you will end up shipping malicious code to customers.

Reviewer 2: ```suggestion
You are likely using Cachito through OSBS. Cachito will serve the malicious dependency to the OSBS build, where you will subsequently use it to build your application container. The malicious behavior may or may not manifest at build time. In the best case scenario, the build will fail. In the worst case scenario, you will end up shipping malicious code to customers.
```

The lock file pins all dependencies to exact versions. For https dependencies, which are impossible to pin, Cachito requires the integrity value (a checksum). For dependencies from the npm registry, Cachito does not require integrity values, but newer versions of npm will always include them. Check if all your dependencies have an integrity value, update your lock file if not.

Do not try to use private registries. If you point npm to a private registry (or any registry other than the official one) via [.npmrc](https://docs.npmjs.com/cli/v6/configuring-npm/npmrc), Cachito will ignore it and look in the official registry anyway.

## yarn

*TL;DR: same as npm, but the handling of unofficial registries is slightly less surprising. Still, you probably should not use them.*

Yarn packages are identical to npm packages except that they use [yarn.lock](https://classic.yarnpkg.com/en/docs/yarn-lock/) as the lock file. Everything that applies to npm applies to yarn.

> The one difference is the handling of unofficial registries. If you point yarn to an unofficial registry via .npmrc or [.yarnrc](https://classic.yarnpkg.com/en/docs/yarnrc), this will be reflected in the resolved url in the lock file. Cachito will see that the url does not point to the official registry and will treat the dependency as a plain https dependency. If the url is accessible to Cachito, it will download the dependency directly without relying on npm/yarn dependency resolution. That does not necessarily make using unofficial registries a good idea. If the registry is private, your build will either fail or leak internal package names.

Reviewer 3: How to understand "leak internal package names"?

Reviewer 4: Let's say you have something like an `internal-yarn-registry.my-company.com`. In there, you have a private package, `some-private-package`, and you're using this as a dependency for a Cachito request. If Cachito is on the same private network as `internal-yarn-registry.my-company.com`, then it may successfully download `some-private-package`, which will then be present in the request tarball, the request JSON and the content manifest. Depending on what you then do with the finished request (e.g. build a container and a source container), the internal package may end up getting leaked to the outside world.
